crates/control/src/search_suggestor.rs

```diff
@@ -0,0 +1,149 @@
+use std::ops::{Index, IndexMut};
+
+use color_eyre::Result;
+use context_attribute::context;
+use coordinate_systems::{Field, Ground};
+use framework::{AdditionalOutput, MainOutput};
+use linear_algebra::{point, Isometry2, Point2};
+use nalgebra::{clamp, DMatrix};
+use serde::{Deserialize, Serialize};
+use types::{
+    ball_position::{BallPosition, HypotheticalBallPosition},
+    field_dimensions::FieldDimensions,
+    parameters::SearchSuggestorParameters,
+};
+
+#[derive(Deserialize, Serialize)]
+pub struct SearchSuggestor {
+    heatmap: Heatmap,
+}
+
+#[context]
+pub struct CreationContext {
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+    search_suggestor_configuration: Parameter<SearchSuggestorParameters, "search_suggestor">,
+}
+
+#[context]
+pub struct CycleContext {
+    search_suggestor_configuration: Parameter<SearchSuggestorParameters, "search_suggestor">,
+    ball_position: Input<Option<BallPosition<Ground>>, "ball_position?">,
+    invalid_ball_positions: Input<Vec<HypotheticalBallPosition<Ground>>, "invalid_ball_positions">,
+    ground_to_field: Input<Option<Isometry2<Ground, Field>>, "ground_to_field?">,
+    heatmap: AdditionalOutput<DMatrix<f32>, "ball_search_heatmap">,
+    field_dimensions: Parameter<FieldDimensions, "field_dimensions">,
+}
+
+#[context]
+#[derive(Default)]
+pub struct MainOutputs {
+    pub suggested_search_position: MainOutput<Option<Point2<Field>>>,
+}
+
+impl SearchSuggestor {
+    pub fn new(context: CreationContext) -> Result<Self> {
+        let (heatmap_length, heatmap_width) = (
+            (context.field_dimensions.length
+                * context.search_suggestor_configuration.cells_per_meter)
+                .round() as usize,
+            (context.field_dimensions.width
+                * context.search_suggestor_configuration.cells_per_meter)
+                .round() as usize,
+        );
+        let heatmap = Heatmap {
+            map: DMatrix::from_element(heatmap_length, heatmap_width, 0.0),
+            field_dimensions: context.field_dimensions.clone(),
+            cells_per_meter: context.search_suggestor_configuration.cells_per_meter,
+        };
+        Ok(Self { heatmap })
+    }
+
+    pub fn cycle(&mut self, mut context: CycleContext) -> Result<MainOutputs> {
+        self.update_heatmap(
+            context.ball_position,
+            context.invalid_ball_positions,
+            context.ground_to_field.copied(),
+            context.search_suggestor_configuration.heatmap_decay_factor,
+        );
+        let maximum_heat_heatmap_position = self.heatmap.map.iamax_full();
```

Comment:
extract to function in `HeatMap` struct, should also do the conversion and return `Point<Field>`

Reply:
Done